### PR TITLE
feat(#23): Pass in a custom fetching function rather than a URL

### PR DESCRIPTION
**Pull request by doug-wade**
I added support for both async functions and promises as the first argument, rather than a url. You can find an example usage [here](https://github.com/doug-wade/i-want-to-be-a-fan/pull/1/commits/9421a2c7be928b0461bf325373858eef3098dbf8#diff-c306e0a99961a16f5c5c83996caa0958b94006d97f97475049ea3a08036bb5b0R40).

Addresses #23

### CODE REVIEW

File: eleventy-fetch.js [hunk 1]
```diff
@@ -34,11 +34,19 @@ function isFullUrl(url) {
 	}
 }
 
+function isAwaitable(maybeAwaitable) {
```
Comment (doug-wade) [inline eleventy-fetch.js:37]: I use this util in a couple of places. We could add a utils file that can be reused, instead, if we'd prefer.

File: eleventy-fetch.js [hunk 2]
```diff
@@ -34,11 +34,19 @@ function isFullUrl(url) {
...
+}
+
 async function save(source, options) {
-	if(!isFullUrl(source)) {
+	if(!(isFullUrl(source) || isAwaitable(source))) {
 		return Promise.reject(new Error("Caching an already local asset is not yet supported."));
 	}
 
+	if (isAwaitable(source) && !options.formatUrlForDisplay) {
+		return Promise.reject(new Error("formatUrlForDisplay must be implemented, as a Promise has been provided."));
```
Comment (doug-wade) [inline eleventy-fetch.js:47]: I'm reusing the `formatUrlForDisplay` function to create a unique cache key. I'm not entirely certain this is the most intuitive name for what this is doing; maybe I should add a send option like `uniqueKey` to the options object instead?

**Comment by zachleat** [comment]
Merged, thank you!

**Comment by zachleat** [comment]
Just info: minor issue with this one, fixed in #52!

**Pull request merged**